icefall/context_graph.py

```diff
@@ -14,243 +14,209 @@
-        kaldifst.arcsort(self.graph)
-
-    def is_final_state(self, state_id: int) -> bool:
-        return self.graph.final(state_id) == kaldifst.TropicalWeight.one
-
-
-    def get_next_state(self, state_id: int, label: int) -> Tuple[int, float, bool]:
-        arc_iter = kaldifst.ArcIterator(self.graph, state_id)
-        num_arcs = self.graph.num_arcs(state_id)
-
-        # The LM is arc sorted by ilabel, so we use binary search below.
-        left = 0
-        right = num_arcs - 1
-        while left <= right:
-            mid = (left + right) // 2
-            arc_iter.seek(mid)
-            arc = arc_iter.value
-            if arc.ilabel < label:
-                left = mid + 1
-            elif arc.ilabel > label:
-                right = mid - 1
-            else:
-                return (arc.nextstate, arc.weight.value, True)
-
-        # Backoff to state 0 with the score on epsilon arc (ilabel == 0)
-        arc_iter.seek(0)
-        arc = arc_iter.value
-        if arc.ilabel == 0:
-            return (0, 0, False)
-        else:
-            return (0, arc.weight.value, False)
+                node = node.next[token]
+            self._fill_fail()
 
+    def forward_one_step(
+        self, state: ContextState, token: int
+    ) -> Tuple[float, ContextState]:
+        """Search the graph with given state and token.
 
-class ContextState:
-    def __init__(self, graph: ContextGraph, max_states: int):
-        self.graph = graph
-        self.max_states = max_states
-        # [(total score, (score, state_id))]
-        self.states: List[Tuple[float, Tuple[float, int]]] = []
-
-    def __str__(self):
-        return ";".join([str(state) for state in self.states])
-
-    def clone(self):
-        new_context_state = ContextState(graph=self.graph, max_states=self.max_states)
-        new_context_state.states = self.states[:]
-        return new_context_state
-
-    def finalize(self) -> float:
-        new_context_state = ContextState(graph=self.graph, max_states=self.max_states)
-        if len(self.states) == 0:
-            return 0, new_context_state
-        item = heappop(self.states)
-        return item[0], new_context_state
-
-    def forward_one_step(self, label: int) -> float:
-        states = self.states[:]
-        new_states = []
-        # expand current label from state state
-        status = self.graph.get_next_state(0, label)
-        if status[2]:
-            heappush(new_states, (-status[1], (status[1], status[0])))
+        Args:
+          state:
+            The given state (trie node) to start.
+          token:
+            The given token.
+
+        Returns:
+          Return a tuple of score and next state.
+        """
+        # token matched
+        if token in state.next:
+            node = state.next[token]
+            score = node.score
+            # if the matched node is the end of a word/phrase, we will start
+            # from the root for next token.
+            if node.is_end:
```

Comment:
`return self.root` only happens when `node` is a leaf, here should be `if node.is_end and node.next == {}` ?
otherwise, if there are two phrases `周杰` and `周杰伦`, it seems `周杰伦` can not be boost ?